### DISCUSSION TITLE
Validate (Sync) API input and return errors instead of returning/relying on SQL query errors

### DISCUSSION THREAD

**Discussion by StevendeVries** [shopware/platform#3157] [RFC / Proposal]
**Effort**: high

### Description:
We use an ESB to connect all kinds of systems with Shopware, at times the data is less then perfect. When sending invalid data to an endpoint you will most likely see an SQL error in the response (with a DriverException/WriteException). Instead of relaying on the database to validate the data send, I would prefer an actual validator and returning a proper exception with an clear error message. The exceptions now are generic is used for a wide range of errors. For instance a WriteException will also occur on a deadlock.

### Benefits
Extending Shopware or plain maintenance will be more cost efficient due to not having the unravel a query compared to a plain error message. And a system like an ESB could do a follow up action with the new response.

### Break strategy
Hmm. The response will be different, changing from a 500 Server error with a DriverException into a 400 (bad request) with a clear error message. Might not actually break compatibility but could for instance impact someone's monitoring rules or behavior acting on the response.

**Comment by OliverSkroblin** [comment]
Hi @StevendeVries,

Thank you for submitting a proposal. To make it a bit more solid, I would ask you to be a bit more specific about what validations you have in mind. So far, we are of course validating the input for the fields for type safety. Deadlocks could be o course be wrapped in a separate excveption. But this is a matter of 2min and not worth a big proposal. 😉

**Comment by StevendeVries** [comment]
Apologies for the long wait, unfortunately been out of commision for a while due to corona :angry: 
Perhaps to make it more solid I looked at our Sentry and grabbed te last error as an example. When updating products using the sync API you might end up with a SQL error like this:
`Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException`

> Uncaught PHP Exception Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException: "An exception occurred while executing 'INSERT INTO `product` (`id`, `version_id`, `parent_version_id`, `product_manufacture.......

> SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails

Instead of relying on the DB structure to catch these, I would prefer a validator checking the data before and if possible make it expandable. When you create a custom entity, you might want to add specific validations. For instance a stringfield having a specific pattern

**Comment by keulinho** [comment]
Thx @StevendeVries  for your proposal, but actually there was not much movement around this in the last year(s) and we don't expect that to change, therefore to be completely transparent we'll close this proposal for now.
